### ISSUE TITLE
Build index in unormalized inner product with lower performance?

### ISSUE THREAD

**Issue by fcjy** [nmslib/hnswlib#74]
Hi @yurymalkov & @searchivarius, 

I have a dataset want to use hnswlib in inner product space(not cosine), it's dimension is 100. I build a index(M = 64, efc = 5000, efs = 5000) and try to do ANN search, recall rate is about 49% and cost 10ms for once search.
Then I try to build index with normalized(M = 64, efc = 5000, efs = 5000), like as cosine space(but actually I want to use ip), and call recall rate with a bruteforce result from ip space, recall rate is about 91% and cost 4ms for once search.
I am confuesd about the unormalized index peforcemance，have you encountered this situation and have any suggestions?
BTW, I also tried faiss::IndexIVFFlat index, it's recall rate is 91%, cost 7ms for once search.

Thanks，
Lin

**Comment by searchivarius** [comment]
@fcjy do you have the same issue with NMSLIB implementation? 
PS: HNSW doesn't always work. I have an example where it doesn't work with the cosine similarity and with the inner product.

**Comment by fcjy** [comment]
> @fcjy do you have the same issue with NMSLIB implementation?
> PS: HNSW doesn't always work. I have an example where it doesn't work with the cosine similarity and with the inner product.

Thank you for your reply!
I had tried NMSLIB, it is better than hnswlib.  it's recall rate is 72%, cost 7ms for once search, with the same parameters.
Maybe faiss::IndexIVFFlat is more suitable for my situation.

**Comment by yurymalkov** [comment]
@fcjy Can you please share a dataset with the problem?
Also, did you tried post=2 in nmslib's HNSW and setting ```EXTEND_USE_EXTENDED_NEIGHB_AT_CONSTR``` to  1 in nmslib's code https://github.com/nmslib/nmslib/blob/master/similarity_search/src/method/hnsw.cc#L52?

I have some ideas why this might happen. One way to solve the problems is using https://github.com/stanis-morozov/ip-nsw code/ (C++), an extension of HNSW done specifically for inner product space.

**Comment by searchivarius** [comment]
@yurymalkov How is that implementation different from HNSW?

**Comment by fcjy** [comment]
@yurymalkov Of course! I had sent an email to you(yurymalkov_mail_ru), please check it.
And I will try the two methods you mentioned today.
Thanks!

**Comment by yurymalkov** [comment]
@fcjy Thank you! I'll test it in a few days.

**Comment by fcjy** [comment]
I do a transform, let the MIPS problem change to NNS problem, such as chapter 2.1 in [this paper](https://papers.nips.cc/paper/7722-non-metric-similarity-graphs-for-maximum-inner-product-search.pdf), and then run libhnsw, that get a very nice result!

**Comment by searchivarius** [comment]
@fcjy could you share your data with me to? (leo at boytsov.info). A one to 10 million sample will be enough. Many thanks!

**Comment by fcjy** [comment]
@searchivarius Of course! I had sent it.

**Comment by wuwenjunwwj** [comment]
hi,  I have encountered the same problem。I have a dataset with 128 dimension and I build a index(M = 200, ef_construction = 1000)  in inner product space，search topK = 5000,  it's call rate is  70%。but when I build index with normalized vector with the same parametor，it‘s rate is about 98% 。Do you know the reason now and have any suggestions ？

**Comment by searchivarius** [comment]
@wuwenjunwwj the inner product search is likely more challenging. It's not the same search problem. Plus, for topK=5000 it's difficult to get accurate results with k-NN search overall.

**Comment by searchivarius** [comment]
@fcjy do you still have that old data set? I tried to download it recently, but clearly it's not there any more :-)
